### ISSUE TITLE
[Package Request] Plex Media Player

### ISSUE THREAD

**Issue by ProfessorKaos64** [ProfessorKaos64/SteamOS-Tools#135]
This application is currently built to `brewmaster_testing` and will stay there for the foreseeable future. This is due to the high version / library requirements of packages, such as QT 5.6.x and more. The application does launch, and it will connect to my existing Plex account.

Currently seeking testers, as this requires a PlexPass, which I currently do not have. I believe at some point, this will be opened up to the public without this requirement.

Preferred method of install (warning, this also may upgrade packages you have, such as Kodi). It is highly suggested you test this on a VM or test machine.

TODO:

- [x] Investigate
- [x] build dependencies (holy mountain of qt5 !!!)
- [x] build package
- [x] testing
- [x] deployment to brewmaster_testing
- [x] Gather feedback from PlexPass subscribers
- [ ] deployment to brewmaster (likely never to happen anytime soon)

### Package info

Current version:  1.1.1.293+git+bsos-1
Updated: 20160525

### Automatic installation

```
cd SteamOS-Tools
./configure-repos.sh --enable-testing
sudo apt-get install plex-media-player
```

### Manual installation

* [individual package](http://packages.libregeek.org/SteamOS-Tools/pool/main/p/plex-media-player/)
* [Package dependencies (See "Depends:") ](https://github.com/ProfessorKaos64/LibreGeek-Packaging/blob/brewmaster/plexmediaplayer/debian/control)

**Comment by Enigma0** [comment]
Hoho am I glad I found this. Hows it going? What kind of tests need doing?

Any quick way to verify that I am definitely running brewmaster_testing. As I understand it the Steam client beta may be a different channel altogether?

**Comment by ProfessorKaos64** [comment]
Away hiking today, but I will provide all this info when I get back. 

On June 3, 2016 11:17:38 AM EDT, Enigma <notifications@github.com> wrote:
>Hoho am I glad I found this. Hows it going? What kind of tests need
>doing?
>
>Any quick way to verify that I am definitely running
>brewmaster_testing. As I understand it the Steam client beta may be a
>different channel altogether?
>
>---
>You are receiving this because you were assigned.
>Reply to this email directly or view it on GitHub:
>https://github.com/ProfessorKaos64/SteamOS-Tools/issues/135#issuecomment-223607873

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by ProfessorKaos64** [comment]
@Enigma0, you would need to add the testing repo. See the main readme for directions. "brewmaster_testing" is a distribution pool of my repository packages.libregeek.org, not  the client beta source, repo.steampowered.com. 

If you want to verify, simply check for the existence of `steamos-tools-beta-repo` in `/etc/apt/sources.list.d`. You can also check the output of `apt-cache policy plex-media-player` as well. 

Be warned, the testing repository will upgrade some existing applications (such as Kodi) if you have them. So it is best to do a backup first, if you have not done one in a while. The disclaimer readme has some details on this. In the readme is also details of how to reference the package lists of, for instance, brewmaster_testing. Annnnnd, the wiki is a great resource as well :)

**Comment by Enigma0** [comment]
I may have installed things non-optimally...

1. There is no SteamOS-Tools directory in presumably the desktop home. (I just have the` configure-repo.sh` script sitting there)
2. I get a `"Invalid option! Exiting..."` response from running `./configure-repos.sh --enable-testing`
3. I don't have a `/etc/apt/sources.list.d` file but a path `/etc/apt/sources.list.d/` which is probably just something I should know I guess. Looks like there isn't any steamos-tools-beta-repo in there though.

Guess I'll get to reading the readme/wiki.

**Comment by ProfessorKaos64** [comment]
You should be able to redownload the script again. The testing option changed 2 or so weeks ago. Cloning the repo is probably best so you can update easily. 

On June 3, 2016 6:27:21 PM EDT, Enigma <notifications@github.com> wrote:
>I may have installed things non-optimally...
>
>1. There is no SteamOS-Tools directory in presumably the desktop home.
>(I just have the` configure-repo.sh` script sitting there)
>2. I get a `"Invalid option! Exiting..."` response from running
>`./configure-repos.sh --enable-testing`
>3. I don't have a `/etc/apt/sources.list.d` file but a path
>`/etc/apt/sources.list.d/` which is probably just something I should
>know I guess. Looks like there isn't any steamos-tools-beta-repo in
>there though.
>
>Guess I'll get to reading the readme/wiki.
>
>---
>You are receiving this because you were assigned.
>Reply to this email directly or view it on GitHub:
>https://github.com/ProfessorKaos64/SteamOS-Tools/issues/135#issuecomment-223709614

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by Enigma0** [comment]
Clone the repo huh. Noobs like me frequently forget how to do that without a supplied command line argument... :)

But ya I've had the script for over a month so that's probably it.

**Comment by Enigma0** [comment]
Tried with the lastest script and got:
```
./configure-repos.sh: line 1: syntax error near unexpected token `<'
./configure-repos.sh: line 1: `<html><body>You are being <a href="https://raw.githubusercontent.com/ProfessorKaos64/SteamOS-Tools/brewmaster/configure-repos.sh">redirected</a>.</body></html>'

```

**Comment by Enigma0** [comment]
Got the brewmaster_testing repo set up but now I am getting this:
```
desktop@steamos:~$ sudo apt-get install plex-media-player 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 plex-media-player : Depends: libmpv1 (>= 0.15) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
desktop@steamos:~$ sudo apt-get install libmpv1 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 libmpv1 : Depends: libass5 (>= 0.13.2) but 0.10.2-3+bsos3 is to be installed
           Depends: libavcodec-ffmpeg56 (>= 7:2.4) but it is not going to be installed
           Depends: libavdevice-ffmpeg56 (>= 7:2.4) but it is not going to be installed
           Depends: libavfilter-ffmpeg5 (>= 7:2.4) but it is not going to be installed
           Depends: libavformat-ffmpeg56 (>= 7:2.4) but it is not going to be installed
           Recommends: youtube-dl (>= 2014.11.26) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
```

**Comment by ProfessorKaos64** [comment]
If you set up your own pinning, undo that. The configure-repos.sh script sets up everything, including this. The values you added are probably conflicting.

**Comment by Enigma0** [comment]
@ProfessorKaos64 Ya fixed that I'm pretty sure. Check the bottom of my previous post. Not sure the bug I hit is necessarily fixable in the short term but maybe. Sounds like maybe ffmpeg needs an update that hasn't yet come which may mean libmpv1 also needs an update?

**Comment by ProfessorKaos64** [comment]
There was probably no need to update libmpv for the purposes of this experiment here. Looking into these dependency issues now.

**Comment by ProfessorKaos64** [comment]
@Enigma0 , thank you for the reminder about ffmpeg. I rebuilt this and the package was installable again. I did a system restore to ensure nothing else was installed. Can you update and try again?

**Comment by Enigma0** [comment]
Awesome! I'll give it a try and report back.

**Comment by Enigma0** [comment]
@ProfessorKaos64 So far so good. Basic functionality is all there.

Notes (I'll add as they become prevalent unless you want me to make individual reports - lemme know):

- Back/exit missing the exit/shutdown/etc options which may is probably just Plex for Linux bug (Relaunch doesn't relaunch - fine for desktop but in SteamOS the UI hangs with loading bars and must be restarted or wait about 1-2 minutes)
- Mouse is invisible until clicked and vanishes frequently while not playing a video (also may just be a Plex for Linux bug in general or SteamOS)

**Comment by ProfessorKaos64** [comment]
Thanks for the feedback! I still wouldn't have a great idea if the run-time dependencies would cause issue with the rest of SteamOS, if this package was in brewmaster (my repository). Does gamepad support work? I need to get on getting a PlexPass subscription. 

>Mouse is invisible until clicked and vanishes frequently

This is the same if you were to launch Google Music player I have or use Chrome. I believe that is a compositor oddity. You could recheck this by launching it in desktop mode. That is not using steamcompmgr/steamos-session, rather gnome-session. 

No it's ok to leave testing data here. Since you have both brewmaster_testing and brewmaster. If you do keep this machine or a side machine loaded with those source lists, it could be a long term indicator if this package could be pushed to brewmaster. I have it on my test machine, but i often reload that one for testing things.

Thanks!

**Comment by Enigma0** [comment]
Updated. Gamepads appear to work (I'lll still have to try PS3 natively with BT but probably does)

Yeah the mouse is just wonky in BPM - works normally in desktop mode.

If you can shell it out, the $150 for lifetime is pretty stellar. No subscription and pays itself off eventually. I gotta say I am really liking the PMP/SteamOS situation. I can leave PMP running in desktop mode and use the steam button to jump back into BPM. Exit to desktop and PMP is right there. Also added it to favorites in BPM but exiting PMP is too timely atm. Takes 1-2 minutes or a restart until you can use BPM again.

**Comment by ProfessorKaos64** [comment]
> Yeah the mouse is just wonky in BPM - works normally in desktop mode.

This has to do with the applications + the compositor bahavior. It seems applications built with/using nodejs/JSON have this kind of behavior, as well as electron-based applications. Perhaps a service is not being closed in a timely manner. If I pickup a sub at some point, I'll see if that goes anywhere. At the very least, I'll continue to update the package here during normal packaging rounds I make.

**Comment by Enigma0** [comment]
Sounds good.

In terms of the service not exiting, I can't actually select an "Exit" action. Instead I can only select "Relaunch" from the exit menu in PMP. So it may be attempting to relaunch for a while and then timing out. I suppose I should just use the BPM overlay to kill it instead.

**Comment by ProfessorKaos64** [comment]
hmm..I'll have to sign up then, since I'm curious now ha. I'll toss on a month sub just to test around for a bit. I don't really want to lay down 150$, a bit much right now for me. Maybe I should throw up a kickstarter! haha

**Comment by Enigma0** [comment]
Ha. After a few weeks on free I was sold. Moving from Kodi was such a bump in ease of use/compatibility. Still use Kodi for some addons though.

In terms of updates, does SteamOS actually run `sudo apt-get update && sudo apt-get upgrade && sudo apt-get dist-upgrade` or is it doing something special? Would be a nice-to-have if SteamOS just kept everything updated.

**Comment by ProfessorKaos64** [comment]
The script `/usr/bin/steamos-update` details this process. Updates are processed via that script and preferences set in unattended upgrades lists under /etc/apt/preferences.d/. My packages are configured to also automatically update. The wiki > [troubleshooting](https://github.com/ProfessorKaos64/SteamOS-Tools/wiki/Troubleshooting#debugging-steamos-upgrade-issues) section notes steps when this process has a hiccup, which is hopefully rarely/never.

So in short, packages are automatically upgraded from the SteamOS-Tools repo, in which a "OS update flag" is displayed in the Steam  client ( I think I have this in the readme too).

**Comment by Enigma0** [comment]
Gotcha.

Turns out that BPM overlay doesn't work over PMP. (updated in the "Notes" post)

**Comment by ProfessorKaos64** [comment]
Strange. I should have time to play around with this tonight, now that I subbed to PlexPass.

**Comment by Enigma0** [comment]
Ok so it seems that only the primary controller can control PMP. I bought a small airmote device that seems to get recognized as the primary controller and can't use a PS3 which now always gets registered as secondary.

**Comment by ProfessorKaos64** [comment]
Strange, I was sick for a few days, but I hope to play around with this soon here.

**Comment by Enigma0** [comment]
Did an update today that force-removed plex-media-player.

**Comment by ProfessorKaos64** [comment]
Hmmm, really? It's still there:

```
mikeyd@archboxmtd SteamOS-Tools $ reprepro list brewmaster_testing | grep plex
brewmaster_testing|main|amd64: plex-media-player 1.1.1.293+git+bsos-1
brewmaster_testing|main|source: plex-media-player 1.1.1.293+git+bsos-1
```

Try installing again please and check `apt-cache policy plex-media-player`. Of note, I will be upgrading Qt5 to 5.7.0 (compatible with Plex-Media-Player builds), and a new Plex-Media-Player will follow.

**Comment by Enigma0** [comment]
```
desktop@steamos:~$ sudo apt-get install plex-media-player
Reading package lists... Done
Building dependency tree
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 plex-media-player : Depends: libmpv1 (>= 0.15) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
```

```
desktop@steamos:~$ apt-cache policy plex-media-player
plex-media-player:
  Installed: (none)
  Candidate: 1.1.1.293+git+bsos-1
  Version table:
     1.1.1.293+git+bsos-1 0
        200 http://packages.libregeek.org/SteamOS-Tools/ brewmaster_testing/main amd64 Packages
```

**Comment by ProfessorKaos64** [comment]
Once I get my test machine up again later tonight, i'll see what's going on. What happens when you try to install `libavcodec-ffmpeg56`? I can rebuild mpv to see if that works fine, since it may be tied to the ffmpeg version tightly.

**Comment by Enigma0** [comment]
It seems to install but apparently it's not the right version.

**Comment by ProfessorKaos64** [comment]
@Enigma0 , should be installable again now after an update. I will be updating Qt this weekend. A new PMP version will follow.

**Comment by Enigma0** [comment]
Nice - got it installed. The pause button/function is gone though.

**Comment by ProfessorKaos64** [comment]
hmmm. Strange. well, let's see how the new version goes next week. I'll update once Qt is upgrade and PMP is as well.

**Comment by Enigma0** [comment]
I've been getting frequent playback glitches. The video is just a static glitched image of gray/black squares/rectangles.

Seems to usually happen when playing something after letting PMP idle a bit. I have to "Relaunch" PMP for it to work again.

Probably something in PMP root but not sure.

**Comment by Enigma0** [comment]
Cannot presently install PMP on an updated SteamOS-Tools pull:

```
desktop@steamos:~/SteamOS-Tools$ sudo apt-get install plex-media-player 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 plex-media-player : Depends: libmpv1 (>= 0.15) but it is not going to be installed
                     Depends: libcec-dev but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
```

**Comment by ProfessorKaos64** [comment]
I hope to get back to PMP soon. I have to update to QT 5.7.x for the latest PMP source code. I will try to get to this in the next week.

**Comment by Enigma0** [comment]
Awesome. It's been working quite well since figuring out turning on Deinterlacing fixed playback issues.

**Comment by ProfessorKaos64** [comment]
The following will work:

```
sudo apt-get install libass5=0.13.2+git+bsos1-1
sudo apt-get install plex-media-player
```

This is due to a high pin value on the SteamOS packages, so that none of them are overwritten. This prevents libass from being installed. I am still unsure how I want to go about this. This is proof enough fact that until SteamOS version 3, this will never leave the testing repository. What I **could** do, is just maintain these manual steps on a wiki page. 

Would that work enough for you?

**Comment by Enigma0** [comment]
Yeah that isn't a bad arrangement. Maybe you could add a parameter to the `./configure` that is like `./configure --plex-media-player` or something and it just does the current manual steps.

**Comment by ProfessorKaos64** [comment]
Sure, I'll def consider that. I ran into some trouble updating Qt 5.7.0 (required to continue building master tree), so it may be a week or so until I get back to updating this. You'll have to bear with me until then.

**Comment by Enigma0** [comment]
No complaints here. I am eternally grateful that you have done all this so far. Having PMP in any version is awesome with SteamOS.

**Comment by Enigma0** [comment]
Still having trouble installing:

```
desktop@steamos:~$ sudo apt-get install libass5=0.13.2+git+bsos1-1
[sudo] password for desktop:
Reading package lists... Done
Building dependency tree
Reading state information... Done
libass5 is already the newest version.
0 upgraded, 0 newly installed, 0 to remove and 2 not upgraded.
desktop@steamos:~$ sudo apt-get install plex-media-player
Reading package lists... Done
Building dependency tree
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 plex-media-player : Depends: libmpv1 (>= 0.15) but it is not going to be instal                                     led
                     Depends: libcec-dev but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
desktop@steamos:~$ sudo apt-get install libmpv1
Reading package lists... Done
Building dependency tree
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 libmpv1 : Depends: libsndio6.1 (>= 1.1.0) but it is not installable
           Recommends: youtube-dl (>= 2014.11.26) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
```

**Comment by ProfessorKaos64** [comment]
I just updated libvmpv to version 18, without sndio (not required, PITA as well). I see your point in adding this to a desktop-software.sh hook. Update your lists and make sure you see version 18 of libmpv1 in apt-cache policy.

**Comment by Enigma0** [comment]
That just a `sudo apt-get update` or something specific to `SteamOS-Tools`? I'll take your word on `sndio` though I can't fathom why it would force a depend when it isn't actually required.

_System is powered down now but will check later._

**Comment by Enigma0** [comment]
```
desktop@steamos:~$ apt-cache policy libmpv1
libmpv1:
  Installed: (none)
  Candidate: 0.18.1+git+bsos1-1
  Version table:
     0.18.1+git+bsos1-1 0
        500 http://packages.libregeek.org/steamos-tools/ brewmaster_testing/main amd64 Packages
     0.17.0+git+bsos2-2 0
        100 /var/lib/dpkg/status
     0.16.0+git+bsos1-1 0
        120 http://packages.libregeek.org/steamos-tools/ brewmaster/main amd64 Packages
     0.6.2-2 0
        110 http://ftp.us.debian.org/debian/ jessie/main amd64 Packages
        110 http://httpredir.debian.org/debian/ jessie/main amd64 Packages
```

**Comment by ProfessorKaos64** [comment]
This is the depends list for libmpv 0.18 : http://slexy.org/view/s20JpE5TUX, there is not sndio in there, as I configure it without it. You may have to remove/purge packages, not sure. In any case, I hit some crappy snags with qtdeclaritive 5.7.0, so it will be a little bit till I figure out that and get the rest of the required qt stack built/updated.

**Comment by Enigma0** [comment]
Hmm. Ok. Well as you know my system is dead in the water at the moment anyhow pending your possible fixit script so obviously no rush on my end. ^^

May end up reformating at some point though so it would be a fresh clean install. My time with a functioning SteamOS install seems to be increasing. The last two times hit failures in weeks.

**Comment by ProfessorKaos64** [comment]
Woohoo! The last qt 5.7.0 piece, qtwebengine, is building now (11,999+ CXX objects!). I had a headache continuing to shoehorn the git sources for Qt into packages (which was fine before), so this time around, and much easier, I backported the Debian experimental 5.7.0 packages. I'll likely do this from now on. But, hopefully 5.7.0 is the minimum for Plex Media Player for a while now (fingers crossed). What a mountain to cross ha. 

I should have the new Plex Media Player up by tomorrow. I am going to hard link the package version for libmpv, so things just install ok, and put a note in the build files somewhere. Because package names changed for some qt packages, and other objects, you will want to remove packages that conflict if you run into any issues. I'll test this from a fresh SteamOS install, but beyond that, you will need to adjust as necessary. Going forward I hope things are a bit smoother. Once confirmed by you, I am going t close this long-winded thread :)

**Comment by ProfessorKaos64** [comment]
Confirmed working. Had to work through several tough issues this time. You'll want to remove old packages. The package is more appropriately named `plexmediaplayer` now, as per the executable. The package will remove the old one. In a few days I'll close this thread unless there is a technical issue. I do agree Relaunch stinks vs closing the application forcefully. There is a request here about that:

https://github.com/plexinc/plex-media-player/issues/354

**Comment by ProfessorKaos64** [comment]
I see... 

>Error details: Uncaught TypeError: Cannot read property 'webChannelTransport' of undefined
Known bug, will be fixed in the next build. It happens when the UI is somehow reloaded, which happens when signing in. Restarting PMP should workaround the error.

http://forums.plex.tv/discussion/comment/1186081#Comment_1186081

**Comment by ProfessorKaos64** [comment]
Verified things worked well after signing back. Closing.

**Comment by ProfessorKaos64** [comment]
I can set the exit power option in resources/settings/settings_description.json. I'll update the package.
